asf_public_discourse_home_decarbonisation/utils/plotting_utils.py

```diff
@@ -0,0 +1,466 @@
+        all_years, post_count_by_year, on="year", how="left"
+    ).fillna(0)
+
+    plt.figure(figsize=(12, 8))
+    sns.barplot(x="year", y="post_count", data=post_count_by_year, color=color)
+    plt.title(
+        f"Distribution of Posts Over Time (By Year) (total sample size n = {sample_size})"
+    )
+    plt.xlabel("Year")
+    plt.ylabel("Number of Posts")
+    plt.tight_layout()
+    plt.savefig(os.path.join(output_path, "Distribution_of_Posts_By_Year.png"))
+    plt.show()
+
+
+########## 2. Plot of the number of users vs number of posts ###########
+
+
+def plot_users_post_distribution(dataframe, output_path):
+    """
+    Creates and saves a plot showing the distribution of the number of posts vs the number of users.
+
+    Args:
+        dataframe (pandas.DataFrame): The DataFrame containing user post data with a 'username' column.
+        output_path (str): The directory path to save the output plot.
+
+    Returns:
+        None: The function saves the plot as a PNG file and displays it.
+    """
+    # Calculate the post counts per user
+    user_post_counts = dataframe["username"].value_counts()
+    sample_size = user_post_counts.size
+    print("sample size:")
+    print(sample_size)
+    # Count how many users have the same number of posts
+    post_count_distribution = user_post_counts.value_counts().sort_index()
+    plt.figure(figsize=(12, 8))
+    # Apply the Freedman-Diaconis Rule to determine bin size
+    iqr = np.subtract(*np.percentile(post_count_distribution, [75, 25]))
+    multiplier = 2.5  # Increase this for larger bins
+    bin_width_fd = multiplier * iqr * (sample_size ** (-1 / 3))
+    bin_size_fd = math.ceil(
+        (max(post_count_distribution) - min(post_count_distribution)) / bin_width_fd
+    )
+    max_bins = 50  # Adjust as needed
+    bin_size_fd = min(bin_size_fd, max_bins)
+    post_range = max(post_count_distribution.index) - min(post_count_distribution.index)
+    number_of_bins = post_range / bin_size_fd
+    # Create a histogram
+    plt.hist(
+        post_count_distribution.index,
+        bins=bin_size_fd,
+        weights=post_count_distribution.values,
+    )
+    # Set the base size for locator based on the number of bins
+    if number_of_bins > 50:
+        base_size = 200
+    elif number_of_bins > 30:
+        base_size = 100
+    elif number_of_bins > 10:
+        base_size = 50
+    else:
+        base_size = 5
+    loc = ticker.MultipleLocator(
+        base=base_size
+    )  # this locator puts ticks at regular intervals
+    plt.gca().xaxis.set_major_locator(loc)
+
+    # Set the title and labels
+    plt.title("Number of Posts vs Number of Users", fontsize=16)
+    plt.xlabel("Number of Posts", fontsize=14)
+    plt.ylabel("Number of Users", fontsize=14)
+    # Rotate the x-tick labels for better readability if there are many bars
+    plt.xticks(rotation=90)
+    # Improve the x-axis by setting the tick marks to only show integers
+    plt.yticks(fontsize=12)
```

Comment:
Ok I might have a look at what you've done to see this. I also may split up the config file like you have. This will also be useful when we do the standardisation as well.